Support Harfbuzz in Metrics View

iorsh opened the pull request:
This PR introduces [HarfBuzz](https://harfbuzz.github.io/) shaping in Metrics view.

1. The current shaping functionality is preserved and can be invoked from View menu as "Built-in shaper". It's very complex, and I'm pretty sure I didn't manage to fully support everything it provides, so some people might find it better in some aspects.
2. The new shaper comes with some generic infrastructure which can be eventually expanded to support more shapers (e.g. Uniscribe under Windows)
3. The new code is written mostly in C++17. This version seems reasonably conservative to me.
4. Editing of kerning values, bearings, and width is supported.
5. The font is generated for HarfBuzz use every time the Metrics View is opened or shaper engine is switched (in the View menu). This could be slow for large fonts. 

**After:**
![aldhabi-hb](https://github.com/user-attachments/assets/2cc97121-9503-4b21-afb4-b86c8c124feb)
**Before:**
![aldhabi-old](https://github.com/user-attachments/assets/b49b6ba0-d8ea-46c8-ac4f-76eb95139275)
**Menu:**
![hb_menu](https://github.com/user-attachments/assets/e58a54fa-cdd3-4955-9edb-5ef488b35655)

 Fixes #3733, fixes #3837, fixes #4158, fixes #4511, fixes #5343, fixes #5513

iorsh commented:
CI is broken.
- Linux: ubuntu-latest moved from 22 to 24
- Windows - pacman issues again.

I'll take a look.

Yukinoroh commented:
iorsh, for your information HarfBuzz is shutting off all OpenType features if Apple Advanced Typography (AAT) tables are present in a font, even for features unique to OpenType. I would really proceed with caution by incorporating HarfBuzz into FontForge.

khaledhosny commented on fontforgeexe/shapers/harfbuzz.cpp:
```diff
@@ -0,0 +1,349 @@
+        hb_glyph_extents_t extents;
+        hb_bool_t res = hb_font_get_glyph_extents(
+            hb_ttf_font, glyph_info.codepoint, &extents);
+        assert(res);
```
Calculating glyph extents can be slow (especially for CFF fonts), and I don’t see extents used anywhere so it seems wasteful.

khaledhosny commented on fontforgeexe/shapers/harfbuzz.cpp:
```diff
@@ -0,0 +1,349 @@
+
+        // Fill unscaled metrics in font units
+        hb_position_t h_advance =
+            hb_font_get_glyph_h_advance(hb_ttf_font, glyph_info.codepoint);
```
You can get the original width from splinechar, shaping does not change it.

khaledhosny commented on fontforgeexe/shapers/harfbuzz.cpp:
```diff
@@ -0,0 +1,349 @@
+    hb_language_t hb_lang = hb_language_from_string((const char*)lang, -1);
+    hb_buffer_set_language(hb_buffer, hb_lang);
+
+    bool rtl = !u_vec.empty() && isrighttoleft(u_vec[0]);
```
Checking first character is not enough, for example if text starts with punctuation. HarfBuzz has [`hb_buffer_guess_segment_properties`](https://harfbuzz.github.io/harfbuzz-hb-buffer.html#hb-buffer-guess-segment-properties) which works slightly better here (it checks the first character with strong direction).

Ideally, there should be a UI to set the direction, since it is possible the text is entirely composed of unencoded glyphs which will be given PUA encoding and all PUA codepoints has strong LTR direction.

khaledhosny commented on fontforgeexe/shapers/harfbuzz.cpp:
```diff
@@ -0,0 +1,349 @@
+
+    _WriteTTFFont(
+        ttf_file, context_->sf, ff_ttf, NULL, bf_ttf,
+        ttf_flag_otmode | ttf_flag_oldkernmappedonly | ttf_flag_fake_map,
```
I don’t think `ttf_flag_oldkernmappedonly` does anything unless `ttf_flag_oldkern` is also set. But you don’t want old kern table anyway.

khaledhosny commented on fontforgeexe/shapers/harfbuzz.cpp:
```diff
@@ -0,0 +1,349 @@
+    _WriteTTFFont(
+        ttf_file, context_->sf, ff_ttf, NULL, bf_ttf,
+        ttf_flag_otmode | ttf_flag_oldkernmappedonly | ttf_flag_fake_map,
+        context_->get_enc_map(context_->sf), ly_fore);
```
You can probably speed things up by skipping generating tables that HarfBuzz does not use, e.g. glyph outlines (glyf and CFF table) which is usually the larges font tables and the slowest to build.

khaledhosny commented on fontforgeexe/shapers/harfbuzz.cpp:
```diff
@@ -0,0 +1,349 @@
+    hb_buffer_set_direction(hb_buffer, hb_dir);
+
+    // Shape the text
+    hb_shape(hb_ttf_font, hb_buffer, NULL, 0);
```
In general, default features are always applied and selecting/unselecting features on the features list has no effect.

khaledhosny commented:
Selecting builtin shaper always crashes for me.

iorsh commented:
Windows CI fixed with https://github.com/jtanx/fontforgebuilds/pull/23

iorsh commented on fontforgeexe/shapers/harfbuzz.cpp:
```diff
@@ -0,0 +1,349 @@
+    if (rtl) {
+        // HarfBuzz reverses the order of an RTL output buffer
+        std::reverse(glyphs_after_gpos, glyphs_after_gpos + glyph_count);
+        std::reverse(width_deltas.begin(), width_deltas.end());
```
This doesn't work for me. The RTL buffer before the reversing contains glyphs tail-to-head, and `hb_buffer_get_glyph_positions()` returns values with geometrically correct left-to-right advances and offsets.

In the case of mixed base glyphs (positive advance) and marks (positive offset, since the mark comes *before* the glyph), the function `hb_buffer_reverse()` correctly reorders the glyphs, but `hb_buffer_get_glyph_positions()` would just return the same values in reverse order, and this breaks the geometric relation.

khaledhosny replied:
If the current code works for you, then fine. `hb_shape()` returns the glyphs in visual order (that is, for RTL text the buffer is reversed), so that code that is drawing the text wouldn‘t need to worry about the buffer direction. `hb_buffer_reverse()` undoes that. Either case the individual glyph advances will be the same since there are no RTL glyph advances in OpenType.

I don’t know what FontForge expects here, but it is probably some FontForge-specific thing, not HarfBuzz that is being unexpected.

iorsh commented on fontforgeexe/shapers/harfbuzz.cpp:
```diff
@@ -0,0 +1,349 @@
+        // functions.
+        hb_bool_t found =
+            hb_font_get_glyph_name(hb_ttf_font, glyph_info.codepoint,
+                                   glyph_name, sizeof(glyph_name) - 1);
```
It looks to me that FF does enforce unique names, at least this is what I got when trying to rename a glyph to existing name: 
![image](https://github.com/user-attachments/assets/9185bbf8-2825-4d6f-ac68-0b90f4647f8c)
Do you have in mind some specific scenario where this doesn't hold?

khaledhosny replied:
Here is an SFD file with duplicate glyph names (two glyphs both named test), FontForge will open it without any complaint and it will also generate a font with the duplicate glyph name: [test.sfd.zip](https://github.com/user-attachments/files/18550394/test.sfd.zip)

khaledhosny commented on fontforgeexe/shapers/harfbuzz.cpp:
```diff
@@ -69,6 +69,17 @@ HarfBuzzShaper::~HarfBuzzShaper() {
+                             HB_FEATURE_GLOBAL_END};
+        hb_feature_vec.push_back(hb_feat);
+    }
+    return hb_feature_vec;
```
Some work is needed to handle features that are enabled by default. Features like `init`, `medi`, etc. are enabled by default and HarfBuzz applies them selectively based on text analysis, but when enabled manually they will be applied unconditionally which will break their intended use.

What should happen is that if a default feature is selected in the UI, it should not be in the features list, but if it is unselected it should be in the features list with value set to 0 (disable).

HarfBuzz does not have an API that tells which features are enabled by default, so you will have to check the HarfBuzz sources. The list of default features depend on the script, I don’t know how FontForge decides which features to enable by default, but I think you will have to sync that with HarfBuzz so that features pre-selected in the UI match what HarfBuzz does.

Some applications use a tri-state checkbox for features, the default is neither selected or unselected what omits the feature from the features list, and selected set the feature to 1 and unselcted sets it to 0.

skef replied:
The HarfBuzz sources are still the best reference, but here's a quick guide based on recent research on which features are considered enabled by default "in general": https://w3c.github.io/IFT/Overview.html#feature-tag-list

iorsh replied:
I added support for default feature list (copied from https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-subset-input.cc#L82), but couldn't see any differentiation by script. @khaledhosny, is there any additional place for default features per script?

khaledhosny replied:
This is the code of the subsetter, not the shaper. For shaping, start here: https://github.com/harfbuzz/harfbuzz/blob/6d8035a99c279e32183ad063f0de201ef1b2f05c/src/hb-ot-shape.cc#L285, where you see the features that applies to all scripts (some features depend on the direction here).

Per-script features are defined in the respective shapers, here, for example, Arabic shaper:
https://github.com/harfbuzz/harfbuzz/blob/6d8035a99c279e32183ad063f0de201ef1b2f05c/src/hb-ot-shaper-arabic.cc#L184

Which shaper is used for which script(s) is handled here:
https://github.com/harfbuzz/harfbuzz/blob/6d8035a99c279e32183ad063f0de201ef1b2f05c/src/hb-ot-shaper.hh#L176

So features depend on the direction and script.

khaledhosny replied:
Some shaper also *disable* some of the common features, e.g. Hangul:
https://github.com/harfbuzz/harfbuzz/blob/6d8035a99c279e32183ad063f0de201ef1b2f05c/src/hb-ot-shaper-hangul.cc#L67

skef replied:
HarfBuzz kinda-sorta has an API to tell which features are enabled by default, if I'm remembering right. If you initialize subset input object (`hb_subset_create_or_fail`) and then call `hb_subset_input_set(input, HB_SUBSET_SETS_LAYOUT_FEATURE_TAG);` on it, you'll get the `hb_set_t` of features as initialized by HarfBuzz. This is the features HarfBuzz retains by default, which is something like the features it expects it may need by default. I don't necessarily recommend using this API, however, not least because HarfBuzz can be installed without subsetting support.

iorsh replied:
Yes, I initially used the feature set from subsetting, but following @khaledhosny's advice, changed it to an explicit collection of features per script. @khaledhosny also provided a new API (https://github.com/harfbuzz/harfbuzz/pull/5047), which kicks in automatically once FontForge is built with supporting HarfBuzz version (10.3+).

iorsh commented:
> Selecting builtin shaper always crashes for me.

@khaledhosny, I tested it with a few fonts, but couldn't reproduce. Could you please attach a crash stack, or the font file if possible?

khaledhosny commented:
> > Selecting builtin shaper always crashes for me.
> 
> @khaledhosny, I tested it with a few fonts, but couldn't reproduce. Could you please attach a crash stack, or the font file if possible?

It happens with any font, when `ff::shapers::IShaper::~IShaper()` is called, this warning seems to be relevant:

```
/fontforge/fontforgeexe/shapers/shaper_shim.cpp:75:5: warning: delete called on 'ff::shapers::IShaper' that is abstract but has non-virtual destructor [-Wdelete-abstract-non-virtual-dtor]
   75 |     delete shaper;
```

This patch makes the warning and the crash go away:
```diff
diff --git a/fontforgeexe/shapers/i_shaper.hpp b/fontforgeexe/shapers/i_shaper.hpp
index 1884b05c1..f44e6f3cb 100644
--- a/fontforgeexe/shapers/i_shaper.hpp
+++ b/fontforgeexe/shapers/i_shaper.hpp
@@ -36,6 +36,8 @@ namespace ff::shapers {

 class IShaper {
  public:
+    virtual ~IShaper() = default;
+
     // The internal shaper name is non-localizable and serves to identify the
     // shaper in the system.
     virtual const char* name() const = 0;
```

iorsh commented on fontforgeexe/shapers/harfbuzz.cpp:
```diff
@@ -0,0 +1,505 @@
+    } else if (script == "Mymr") {
+        return features_khmer;
+    } else {
+        return features_use;
```
@khaledhosny, I tried to duplicate HarfBuzz logic here, but I'm confused about `_hb_ot_shaper_default` vs `_hb_ot_shaper_use`. I chose the latter by solely considering _the more the merrier_, but I'm not quite sure this is the way to go. Could you please take a look?

khaledhosny replied:
USE (Universal Shaping Engine) is used for most new scripts in Unicode. Default is the default shaper that is used for scripts that do not have a dedicated shaper (e.g. Latin). HarfBuzz might chose default shaper for scripts that should go through USE shaper if the font does not declare a script (i.e. all features are under `dflt` script).

khaledhosny commented:
I’m working on a HarfBuzz API that returns the features enabled on a shape plan, see https://github.com/harfbuzz/harfbuzz/pull/5047. Can you try this API and see if it works for your case? (you can see examples of how to use it in the unit tests).

iorsh commented:
> I’m working on a HarfBuzz API that returns the features enabled on a shape plan, see [harfbuzz/harfbuzz#5047](https://github.com/harfbuzz/harfbuzz/pull/5047). Can you try this API and see if it works for your case? (you can see examples of how to use it in the unit tests).

Sure, I'll check that. I noticed that in unit tests you are supplying a default `" "` buffer, and I'm planning to do the same. As far as I understand, supplying actual buffer could lead to changes in default feature list, and Metric View UI is not built to accommodate them. With the default buffer, the displayed feature list would only depend on direction and script.

khaledhosny commented:
> > I’m working on a HarfBuzz API that returns the features enabled on a shape plan, see [harfbuzz/harfbuzz#5047](https://github.com/harfbuzz/harfbuzz/pull/5047). Can you try this API and see if it works for your case? (you can see examples of how to use it in the unit tests).
> 
> Sure, I'll check that. I noticed that in unit tests you are supplying a default `" "` buffer, and I'm planning to do the same. As far as I understand, supplying actual buffer could lead to changes in default feature list, and Metric View UI is not built to accommodate them. With the default buffer, the displayed feature list would only depend on direction and script.

To be honest, it didn’t occur to me that the text content can change the enabled features. I used an empty string in the unit tests because they are not concerned about the actual shaping.

khaledhosny commented on fontforgeexe/shapers/harfbuzz.cpp:
```diff
@@ -502,4 +505,48 @@ std::set<Tag> HarfBuzzShaper::default_features(Tag script,
+    // dummy call to check the total projected number of tags
+    unsigned int zero = 0;
+    unsigned int count =
+        hb_ot_shape_plan_get_feature_tags(shape_plan, 0, &zero, nullptr);
```
You can pass nullptr instead of the dummy &zero.

iorsh commented:
@khaledhosny, I think I have addressed all your comments. Could you please review again?

khaledhosny commented:
Looks good, thanks! 

A few observations:
* Opening the metrics window still takes a long time for some fonts (e.g. Amiri):
* This happens every time the metrics window is opened even if the font did not change at all. I guess the temporary font is not cached, but may be it should.
* Guessing direction does not still work for unencoded glyphs. If the glyph string is all unencoded glyphs, the direction remains LTR even if I select Arabic script in the UI

khaledhosny commented:
> HarfBuzz does lots of caching internally, though, so you might want to destroy the face and re-create it when one of the tables changes, but you can still cache the unchanged blobs independent of the HarfBuzz face, so creating a new face will be a light operation.

Actually, there is `hb_font_changed()` to invalidate the internal caches.

skef commented:
Ah, right, I forgot `hb_face_create_for_tables`.

Unfortunately I'm not sure the on-demand aspect of that API, driven from within HarfBuzz, will help much. I suppose it could help simplify some of the changes: You keep the overall refresh operations the same, and arrange for `hb_font_changed` to be called whenever the window would be refreshed. Then if you've hooked up the callback for hb_face_create_for_tables to the right generators the problem boils down to determining what is stale, which means you can keep the staleness indicator close to the structure rather than having to signal it back to some more central place.

iorsh commented on fontforge/splinechar.h:
```diff
@@ -0,0 +1,654 @@
+/* Copyright (C) 2000-2012 by George Williams */
```
This is a sort of cleanup, which I might have skipped, but `splinefont.h` has structures with `class` and `new` fields, which are incompatible with C++. I could rename them, but this would have just polluted the PR with a lot of unrelated changes.

skef replied:
Oh, right, that makes particular sense.

khaledhosny commented on fontforgeexe/metricsview.c:
```diff
@@ -3948,6 +3950,7 @@ static ShaperContext* MVMakeShaperContext(MetricsView *mv) {
     context->get_enc_map = SFGetMap;
     context->get_char_width = MVCharWidth;
     context->get_kern_offset = MVGetKernOffset;
+    context->script_is_rtl = ScriptIsRightToLeft;
```
HarfBuzz has `hb_script_get_horizontal_direction()` which might be more up to date (`hb_buffer_guess_segment_properties()` will call it if the direction is not set or set to `HB_DIRECTION_INVALID`). But I still think a separate UI control to set direction is needed since a few scripts can be written horizontally in both directions.

iorsh replied:
ScriptIsRightToLeft() is passed for use in the legacy shaper, HarfBuzz won't use it. Regarding the UI, I have some enhancements in mind, but for now I prefer to keep UI changes at the necessary minimum. GDraw poses quite a few challenges in its workings.

khaledhosny commented on fontforgeexe/shapers/harfbuzz.cpp:
```diff
@@ -132,7 +133,10 @@ SplineChar** HarfBuzzShaper::extract_shaped_data(hb_buffer_t* hb_buffer) {
+        auto ttf_map_it = ttf_map_.find(glyph_info.codepoint);
+        SplineChar* glyph_out = (ttf_map_it != ttf_map_.end())
+                                    ? ttf_map_it->second
+                                    : get_notdef_glyph();
```
There is `hb_buffer_set_not_found_glyph()` that allow setting any arbitrary glyph index to be used instead of 0, you can, say, set it to a special value that is guaranteed not to be a valid glyph index like `INT32_MAX` and intercept it here. In case it makes things any simpler/faster.

iorsh replied:
Yes, I tried it, but Ubuntu 22 has HarfBuzz 2.7.4, and `hb_buffer_set_not_found_glyph()` was introduced in 3.1.0. Besides, I need to check the glyph index anyway, it would be a crash otherwise.

iorsh commented:
The high-level HarfBuzz shaper structure is as follows:

- When opening the "Metrics View" window, the shaper is initialized with TTF file generated on-fly.
- While the window is active, it is responsive to the changes in glyph kerning using `HarfBuzzShaper::initial_kerning_`. The initial kerning is preserved, and once its value changes, the delta is applied internally over the metrics computed by HarfBuzz.
- In a similar manner changes to glyph width are also addressed with `HarfBuzzShaper::initial_width_`, without the need to regenerate the TTF.
- In case of more complex changes, such as in GSUB/GPOS tables, the user is expected to perform manual refresh, and I added "Refresh Shaper" menu item for that purpose. During the refresh the TTF file is fully regenerated.
- In total, the shaper is regenerated whenever
  - A new "Metrics View" is opened
  - The user hits "Refresh Shaper"
  - The uses changes the shaper ("Built-in" to "HarfBuzz" or vice versa)
 
The building of HarfBuzz shaper is heavy, and eventually we shall make an effort to improve its performance. Currently the TTF is written to a temporary file, and as far as I can see, it's not even guaranteed to reside in RAM. In Windows it's always on the disk (see `GFileTmpfile()`), and in Linux it probably depends on the mount method of `/tmp`.

For now I would prefer to keep the scope of this PR as is, but I can see several possibilities to improve the performance in future:
- Refactoring of `_WriteTTFFont()`, so that it actually uses a generic buffer which can be either kept in memory or dumped onto disk.
- To employ the granular approach with `hb_face_create_for_tables()`, it would be necessary to refactor the change tracking. Currently there is a bunch of `changed_???` flags scattered across multiple structures, and it would be good to harmonize them with some unified approach.
- Inherent FF deficiencies, e.g. the names hashtable `struct glyphnamehash` with 257 baskets, which is totally inadequate for fonts with thousands of glyphs.

skef commented:
I'll take another look at this review-wise, but I suggest that @iorsh should start thinking about documentation.

brigazvi commented:
THIS IS HUGE!!!!
@iorsh thank you so much for all the hard work you put into this. for RTL designers like me it's makes such a difference.

in my opinion it justifies a new release.

tuffnatty commented on fontforge/splinefont_enums.h:
```diff
@@ -64,6 +64,7 @@ enum ttf_flags {
     ttf_flag_dummyDSIG         = 1 << 15,
     ttf_native_kern            = 1 << 16, // This applies mostly to U. F. O. right now.
     ttf_flag_fake_map          = 1 << 17, // Set fake unicode mappings for unmapped glyphs
+    ttf_flag_no_outlines       = 1 << 18, // HarfBuzz: Skip "glyf" and "CFF " tables for performance
```
@iorsh Hi! Here you're adding `ttf_flag_no_outlines` which is accidentally equal to `ps_flag_nohints`. So, after this change, a `font.generate()` call from Python with a `('no-hints')` flags passed generates a TTF without outlines, which is quite annoying and took some time to bisect and debug.
Probably `ttf_native_kern` (which is quite old) and `ttf_flag_fake_map` (which belongs to this changeset) which are conflicting with `ps_flag_nohintsubs` and `ps_flag_noflex`, also need to be addressed. And we're running out of bit space, so these values can't be just moved after 1 << 30. Probably `ps_flag_nohints` does not mean much in TTF building context, I am not sure; but changing the existing behaviour so that a corrupted file is generated seems to me a thing that should be avoided.
